### ISSUE TITLE
Vulkan source debugging does not work with remote hosts

### ISSUE THREAD

**Issue by w-pearson** [baldurk/renderdoc#3076]
## Description

It's not possible to use source debugging on Vulkan when connected to a remote host, despite the shader having the relevant debug information.

This affects both android and non-android remote hosts (I've tested with android and a """remote""" host on 127.0.0.1).

The [shader viewer checks for `sourceDebugInformation`](https://github.com/baldurk/renderdoc/blob/0202a066a9ffa4077fa980da2d97ad5fb6ec5827/qrenderdoc/Windows/ShaderViewer.cpp#L571-L575) which is set [in `spirv_reflect.cpp`](https://github.com/baldurk/renderdoc/blob/0202a066a9ffa4077fa980da2d97ad5fb6ec5827/renderdoc/driver/shaders/spirv/spirv_reflect.cpp#L963-L967) based on `knownExtSet[ExtSet_ShaderDbg]`, which in turn is set [by `spirv_processor.cpp`](https://github.com/baldurk/renderdoc/blob/0202a066a9ffa4077fa980da2d97ad5fb6ec5827/renderdoc/driver/shaders/spirv/spirv_processor.cpp#L526-L537) based on it seeing the `NonSemantic.Shader.DebugInfo.100` extension. But it doesn't seem like that code is run at all (by either qrenderdoc or renderdoccmd) when using the remote server. Given that assembly debugging works with the remote server, and all of the needed information for source debugging is present, I'd expect source debugging to work too, but maybe I'm missing a complication?

### Setup

I tested this with SaschaWillems' deferred shadows sample, as that's what's shown in [this talk](https://vulkan.org/user/pages/09.events/vulkanised-2023/vulkanised_2023_source_level_shader_debugging_in_vulkan_with_renderdoc.pdf)/[video](https://www.youtube.com/watch?v=Fja4lT508cA).

1. Clone https://github.com/SaschaWillems/Vulkan
2. Edit `compileshaders.py` to use `--gVS` and to work around another error:

    ```patch
    diff --git a/shaders/glsl/compileshaders.py b/shaders/glsl/compileshaders.py
    index e0a4e1f5..64811547 100644
    --- a/shaders/glsl/compileshaders.py
    +++ b/shaders/glsl/compileshaders.py
    @@ -38,9 +38,9 @@ for root, dirs, files in os.walk(dir_path):
     
                 add_params = ""
                 if args.g:
    -                add_params = "-g"
    +                add_params = "-gVS"
     
    -            if file.endswith(".rgen") or file.endswith(".rchit") or file.endswith(".rmiss"):
    +            if file.endswith(".rgen") or file.endswith(".rchit") or file.endswith(".rmiss") or "rayquery" in root:
                    add_params = add_params + " --target-env vulkan1.2"
     
                 res = subprocess.call("%s -V %s -o %s %s" % (glslang_path, input_file, output_file, add_params), shell=True)
    ```

    (per https://github.com/SaschaWillems/Vulkan/issues/880 there are other ways of building the shaders, but this seems like the easiest one to me)
3. Run `python compileshaders.py --g` to build the shaders with debug info.
4. Follow the instructions in https://github.com/SaschaWillems/Vulkan/blob/master/BUILD.md to build the samples themselves. (I used a debug build.)

### Working local case

1. Open RenderDoc
2. Launch deferredshadows.exe.
3. Make a capture, and then close deferredshadows.exe and let the capture automatically open.
4. Select the `vkCmdDraw` event in the second color pass.
5. Select pipeline state, then "fragment shader", then "view".
6. Switch to the assembly tab, and change the disassembly type to `SPIR-V Asm (spirv-dis)`.
7. Observe that the assembly includes:
    ```
                   OpExtension "SPV_KHR_non_semantic_info"
              %2 = OpExtInstImport "NonSemantic.Shader.DebugInfo.100"
    ```
8. In the texture viewer, select a pixel, and then select "debug".
9. Observe that it defaults to the source tab, and that a "debug in assembly" button is enabled.
10. In the assembly tab, observe that `Extension("SPV_KHR_non_semantic_info");` is present.

### Broken remote case

1. Run `renderdoccmd remoteserver`
2. Open RenderDoc.
3. Select "tools" &rarr; "manage remote servers"
4. In the hostname text field, enter `127.0.0.1`, then click "add". Close the remote servers window.
5. Switch the replay context to 127.0.0.1.
6. Launch deferredshadows.exe.
7. Make a capture, and then close deferredshadows.exe and let the capture automatically open.
8. Select the `vkCmdDraw` event in the second color pass.
9. Select pipeline state, then "fragment shader", then "view".
10. Switch to the assembly tab, and change the disassembly type to `SPIR-V Asm (spirv-dis)`.
11. Observe that the assembly includes:
    ```
                   OpExtension "SPV_KHR_non_semantic_info"
              %2 = OpExtInstImport "NonSemantic.Shader.DebugInfo.100"
    ```
12. Select a pixel, and then select "debug".
13. Observe that "source debugging unavailable" is displayed, although it still defaults to the source tab.
14. In the assembly tab, observe that `Extension("SPV_KHR_non_semantic_info");` is present.

## Environment

* RenderDoc version: v1.29, v1.28
* Operating System: Windows 10
* Graphics API: Vulkan

**Comment by baldurk** [comment]
That commit should fix the problem I found, but I'm not sure I was able to repro the problem you were talking about.

That flag controlled a cosmetic tool button which toggled between source/disassembly, but otherwise everything all worked correctly for source debugging over a remote host. The same functionality was available in the context menu to jump between viewing the currently disassembly instruction and current source line.

But the title and first line of the description says that you found the source debugging doesn't work. If that's the case and you weren't mistaken, and it still happens after that commit, can you please open a new issue with the steps to reproduce the actual failure in debugging you were seeing?

**Comment by w-pearson** [comment]
That commit seems to fix everything as far as I can tell.

I think I must have confused myself a lot, partially by trying to debug it and determining that the check in `spirv_processor.cpp` wasn't running (which is incorrect; it _does_ run in renderdoccmd, which I confirmed in a lazy way with `std::abort()` - I'm not sure how I didn't run into that in my earlier testing).

The other source of confusion was that I had also tested the same deferredshadows sample with `-g` instead of `-gVS`, i.e. with embedded source but not the extra nonsemantic debugging info. In that case both the disassembly and source tabs show up, and the context menu also has "Go to Source"/"Go to Assembly" options, and stepping while on the source tab still steps by lines; the "source debugging unavailable" message is the only obvious distinguishing factor between `-g` and `-gVS`. So I assumed that since the "source debugging unavailable" message was showing up even with `-gVS` when using a remote host, it was behaving as if the nonsemantic info was not being read properly. It sounds like that wasn't the case though.
